### DISCUSSION TITLE
Communicate request body validation errors to the client

### DISCUSSION THREAD

**Discussion by sawa-ko** [loco-rs/loco#987] [Q&A]
I don't know how to report validation errors that have occurred when trying to use a controller, as currently only the message “Bad request” is displayed but this does nothing to help solve the problem of why the request could not be processed.

![image](https://github.com/user-attachments/assets/b2fafe2f-a1c0-4301-b01f-f41712d7500d)

Actually I see that the validation error does appear in the logs of the application, and I was wondering if there is a way to serialize the validation error that appears in the logs so that it is shown in the error message received by the client that uses the controller that gave error.

![image](https://github.com/user-attachments/assets/65b97cd9-04b7-4477-8ea4-f0dec9655baa)

If this error message could be transmitted, the client consuming the controller would know how to correctly send the body expected by the controller.

**Comment by avk458** [comment]
Because the scaffold controller uses the `loco_rs::Result` type alias, the code in [controller.rs](https://github.com/loco-rs/loco/blob/80885e9354c1cdc2c78ebe34741de4699c8e1098/src/controller/mod.rs#L234) turns every other error into a bad request.
You can create an error type and implement `IntoResponse` as you like.

**Comment by CharlieBytesX** [comment]
Hi, does this continue to be the default way of doing things?

**Reply by CharlieBytesX** [reply]
@sawa-ko what approach are you using to send validation errors to user?

**Reply by saurabh-git-dev** [reply]
@CharlieBytesX 
I had the same issues... See here. This should solve the request body and query params validation. Although for the path params, follow the discussion. https://github.com/loco-rs/loco/discussions/1543

### Solution(still lacks a few things)
The `QueryValidateWithMessage` and `JsonValidateWithMessage` are the built-in extractors( merged in the latest version) that handle the validation errors. and returns the actual errors in a proper format.
There are also some limitations.. There might be an error during JSON serialization/deserialization (due to missing fields), which is not handled by default. and still result in just "Bad Request".
To fix this issue, the fields can be made optional, but add a "required" validation attribute to them. This will allow JSON deserialization to succeed without throwing an error. and the validation will check if the value is passed or not. 

The `PathValidateWithMessage` don't exist yet but can be easily created... See here.. https://github.com/Davidzhu001/loco/blob/792efde7b5ed4b45dfbb700320367a9ae73dae41/src/controller/extractor/validate.rs#L221
The person was trying to implement, but there were some limitations, according to him. but for a simple use case, it should work.

### Current Issues
Those extractors solve the issue to some extent. but still, 2 issues are present...
1. If we pass a number to a string pointer(ex, an Optional email field), it will also fail during the deserialization process, which will return early and result in just a "Bad Request" response, not a full response. (Probably we need a custom JSON deserialization method).
Working in Go, I saw similar problems.. There are also no proper ways to handle errors(JSON marshalling/unmarshalling and returning proper errors). So, I wrote my whole custom parsing logic on top of go-validator to parse path, query params and body.

2. The validation error messages are more verbose.. There are no proper in-built solutions to get friendly error messages. (It's a limitation of the validator crate.)

See some results
<img width="1193" height="1229" alt="image" src="https://github.com/user-attachments/assets/fbb38142-0e7d-4247-b6b5-07b4a2befcf8" />
<img width="1149" height="990" alt="image" src="https://github.com/user-attachments/assets/737747b8-78dd-4979-806b-74ceef008dde" />
<img width="1164" height="820" alt="image" src="https://github.com/user-attachments/assets/94a29698-0f00-41a0-bb27-2cc371a105cd" />
